pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,984 @@
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) deleteBGPPolicy(obj interface{}) {
+	bp := obj.(*v1alpha1.BGPPolicy)
+	if !c.matchedCurrentNode(bp) {
+		return
+	}
+	klog.V(2).InfoS("Processing BGPPolicy DELETE event", "BGPPolicy", klog.KObj(bp))
+	c.queue.Add(dummyKey)
+}
+
+func getIngressIPs(svc *corev1.Service) []string {
+	var ips []string
+	for _, ingress := range svc.Status.LoadBalancer.Ingress {
+		if ingress.IP != "" {
+			ips = append(ips, ingress.IP)
+		}
+	}
+	return ips
+}
+
+func (c *Controller) matchedCurrentNode(bp *v1alpha1.BGPPolicy) bool {
+	node, _ := c.nodeLister.Get(c.nodeName)
+	return matchedNode(node, bp)
+}
+
+func matchedNode(node *corev1.Node, bp *v1alpha1.BGPPolicy) bool {
+	nodeSelector, _ := metav1.LabelSelectorAsSelector(&bp.Spec.NodeSelector)
+	return nodeSelector.Matches(labels.Set(node.Labels))
+}
+
+func matchedService(svc *corev1.Service, bp *v1alpha1.BGPPolicy) bool {
+	ipTypeMap := serviceIPTypesToAdvertise(bp.Spec.Advertisements.Service.IPTypes)
+	if ipTypeMap.Has(v1alpha1.ServiceIPTypeClusterIP) && len(svc.Spec.ClusterIPs) != 0 ||
+		ipTypeMap.Has(v1alpha1.ServiceIPTypeExternalIP) && len(svc.Spec.ExternalIPs) != 0 ||
+		ipTypeMap.Has(v1alpha1.ServiceIPTypeLoadBalancerIP) && len(getIngressIPs(svc)) != 0 {
+		return true
+	}
+	return false
+}
+
+func (c *Controller) filterAffectedBPsByService(svc *corev1.Service) sets.Set[string] {
+	affectedBPs := sets.New[string]()
+	allBPs, _ := c.bgpPolicyLister.List(labels.Everything())
+	for _, bp := range allBPs {
+		if !c.matchedCurrentNode(bp) || bp.Spec.Advertisements.Service == nil {
+			continue
+		}
+		if matchedService(svc, bp) {
+			affectedBPs.Insert(bp.GetName())
+		}
+	}
+	return affectedBPs
+}
+
+func (c *Controller) hasAffectedBPsByService(svc *corev1.Service) bool {
+	allBPs, _ := c.bgpPolicyLister.List(labels.Everything())
+	for _, bp := range allBPs {
+		if !c.matchedCurrentNode(bp) || bp.Spec.Advertisements.Service == nil {
+			continue
+		}
+		if matchedService(svc, bp) {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *Controller) addService(obj interface{}) {
+	svc := obj.(*corev1.Service)
+	if c.hasAffectedBPsByService(svc) {
+		klog.V(2).InfoS("Processing Service ADD event", "Service", klog.KObj(svc))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) updateService(oldObj, obj interface{}) {
+	oldSvc := oldObj.(*corev1.Service)
+	svc := obj.(*corev1.Service)
+
+	if slices.Equal(oldSvc.Spec.ClusterIPs, svc.Spec.ClusterIPs) &&
+		slices.Equal(oldSvc.Spec.ExternalIPs, svc.Spec.ExternalIPs) &&
+		slices.Equal(getIngressIPs(oldSvc), getIngressIPs(svc)) &&
+		oldSvc.Spec.ExternalTrafficPolicy == svc.Spec.ExternalTrafficPolicy &&
+		reflect.DeepEqual(oldSvc.Spec.InternalTrafficPolicy, svc.Spec.InternalTrafficPolicy) {
+		return
+	}
+	oldAffectedBPs := c.filterAffectedBPsByService(oldSvc)
+	newAffectedBPs := c.filterAffectedBPsByService(svc)
+	if oldAffectedBPs.Len() != 0 || newAffectedBPs.Len() != 0 {
```

Comment:
I don't understand why it checks the number of affected policies. Shouldn't it just check if any of the old and new policy has affected policies?

Reply:
I should have updated this. This is the work-around code between the refactoring the multiple workers to single worker to pass the unit tests.